specs/logging_implementation.adoc

```diff
@@ -0,0 +1,274 @@
+
+== Proposed change
+* Logs and alerts need to be treated as messages emitted by Tendrl modules.
+* A schema needs to be defined for emitted messages.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must pass the message as per schema.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+None
+
+== Impacted modules
+
+=== Tendrl/common impact
+* Create a new class called "Message" in common for message handling with attributes
+  message_type, filename, function_name, line_number, timestamp.
+* Create a new class called "Log" in common with attributes log_level, message,
+  log_trace_back. Inherit Log class from Message class.
+* Create a new class called "Alert" in common with attributes alert_id, node_id,
+  resource, current_value, tags, type, severity, significance, ackedby, acked.
+  Inherit alert class from Message class. (loglevel for alert is always Warning, no need
+  any log information for alert because everything is predefined, so no need any log class help
+  for alert).
+* Create a new class called "FlightOperation" in common with attributes
+  flight_operation_id, message. Inherit FlightOperation class from Message class.
+```
+Log class object structure
+   - log_level
+   - message
+   - log_trace_back (True/False)
+
+Alert class object structure
+  - alert_id
+  - node_id
+  - resource
+  - current_value
+  - tags
+  - type
+  - severity
+  - significance
+  - ackedby
+  - acked
+
+Flight operation class object structure
+  - flight_operation_id
+  - message
+
+Message class object structure
+  - log_handler_name
+  - message_type
+  - filename
+  - function_name
+  - line_number
+  - timestamp
+```
+* Create a new function called message_handler in Message class and process the
+  object values.
+* If the message type is Alert then store the message in log file with uuid (log level
+  for alert is always warning). and put the object in node_agent socket.
```

Comment:
i have discussed with @shtripat ,  @anmolbabu and @nnDarshan. Alert needs different log levels.